### PR TITLE
RTL8139: Alloc registers resource and perform software reset

### PR DESCRIPTION
**Pull request by komik0**
With this commit the following changes are introduced:

- resource for rtl8139 regs is added
- software reset is performed

### CODE REVIEW

File: sys/drv/rtl8139.c
Comment (pj1031999) [inline sys/drv/rtl8139.c:33]: You can write
`for (int i = 0; i < RL_TIMEOUT; i++)`

File: sys/drv/rtl8139.c
Comment (pj1031999) [inline sys/drv/rtl8139.c:48]: I don't know how that device works but maybe we should return some error here in case of timeout?

File: sys/mips/gt64120.c
Comment (pj1031999) [inline sys/mips/gt64120.c:350]: Please consult that with @mohrcore, @ThomasStachovsky, @MichalBlk or @cahirwpz.

Comment (MichalBlk) [reply sys/mips/gt64120.c:350]: This PR is blocked until merging #826 .

**Comment by MichalBlk** [comment]
Please adjust to #826

File: sys/kern/kmem.c
Comment (pj1031999) [inline sys/kern/kmem.c:110]: Why you removed `page_aligned_p` from `assert`?

File: sys/drv/rtl8139.c
Comment (MichalBlk) [inline sys/drv/rtl8139.c:10]: The `rtl8139_state` structure absolutely can't contain an `intr_handler_t` structure. Please note that interrupt handlers are only created inside the interrupt module and are allocated form the `M_INTR` kmalloc pool. Please remove this field. If you ever need an irq, you should use the `resource_t` structure.

**Review by MichalBlk — CHANGES_REQUESTED** [review]

File: sys/drv/rtl8139.c
Comment (MichalBlk) [reply sys/drv/rtl8139.c:28]: Ok, for now there is no good way to wait without doing busy waiting.

File: sys/drv/rtl8139.c
Comment (komik0) [reply sys/drv/rtl8139.c]: ok, it is just outdated

**Review by komik0 — COMMENTED** [review]
cs is ready for review

File: sys/drv/rtl8139.c
Comment (MichalBlk) [reply sys/drv/rtl8139.c:43]: It's interesting. I presume that reset initializes some internal state (like counters and registers). If it fails then what guarantee do you have about consistency of the state?
It would be nice to add a comment which explains that this kind of failure isn't fatal.

File: include/dev/rtl8139_reg.h
Comment (MichalBlk) [reply include/dev/rtl8139_reg.h:19]: Why not just `RL_CFG1` and `RL_CFG1_...`? Do you really want to literally copy all definitions from their header file? Maybe they have other devices starting with prefix "RL" and hence they use `RF_8139_CFG1` instead of `RF_CFG1`.

File: include/dev/rtl8139_reg.h
Comment (MichalBlk) [inline include/dev/rtl8139_reg.h:7]: The layout of this header file is strange. You labeled the definition of `RL_CMD_RESET` with notable comment
```C
/*
 * Command register.
 */
```
but totally separated it from the definition of `RL_COMMAND`. Why not
```C
/*
 * Command register.
 */
#define RL_COMMAND 0x0037
#define RL_CMD_RESET 0x0010
```

File: include/dev/rtl8139_reg.h
Comment (MichalBlk) [inline include/dev/rtl8139_reg.h:9]: The same thing, why not
```C
/*
 * Config register #1
 */
#define RL_CFG1 0x0052
#define RL_CFG1_PWRON 0x00
```

File: include/dev/rtl8139_reg.h
Comment (MichalBlk) [reply include/dev/rtl8139_reg.h:21]: I agree. Some bits require a few writes to actually obtain some effect. This driver isn't the only one which uses this approach and the value (i.e. `TIMEOUT`) is most probably magic.

File: include/dev/rtl8139_reg.h
Comment (MichalBlk) [inline include/dev/rtl8139_reg.h:19]: I guess that this define describes some particular setting of the first configuration register (probably "power on"?).
It would be nice to add a brief comment beside the definition, like this
```C
#define  RL_CFG1_PWRON 0x00 /* power on */
```

File: sys/kern/kmem.c
Comment (MichalBlk) [reply sys/kern/kmem.c]: It absolutely doesn't and you can't remove this assert. I think that this assert must be here and is essential form the perspective of the `kmem_map` function. We need to change logic in bus driver (probably in activate function).

File: include/dev/rtl8139_reg.h
Comment (komik0) [reply include/dev/rtl8139_reg.h:7]: as stated in comment - it is copied from freebsd - I only removed not needed constants.
I think in such cases it is expected to do not reorder registers - that approach is useful if we want to use difftool in future.

File: include/dev/rtl8139_reg.h
Comment (komik0) [reply include/dev/rtl8139_reg.h]: It was idea to copy whole file and do not introduce changes wherever possible

Comment (komik0) [reply include/dev/rtl8139_reg.h]: done

File: include/dev/rtl8139_reg.h
Comment (komik0) [reply include/dev/rtl8139_reg.h:9]: as above

**Review by MichalBlk — APPROVED** [review]

**Review by cahirwpz — APPROVED** [review]
LGTM after some fixes.

**Pull request merged**
